src/flp/gadgets.rs

```diff
@@ -438,25 +464,43 @@ where
     fn call_poly(&mut self, outp: &mut [F], inp: &[Vec<F>]) -> Result<(), FlpError> {
         gadget_call_poly_check(self, outp, inp)?;
 
+        // Create a copy of the inner gadget and two working buffers on each thread. Evaluate the
+        // gadget on each input polynomial, using the first temporary buffer as an output buffer.
+        // Then accumulate that result into the second temporary buffer, which acts as a running
+        // sum. Then, discard everything but the partial sums, add them, and finally copy the sum
+        // to the output parameter. This is equivalent to the single threaded calculation in
+        // ParallelSum, since we only rearrange additions, and field addition is associative.
         let res = inp
             .par_chunks(self.serial_sum.inner.arity())
-            .map(|chunk| {
-                let mut inner = self.serial_sum.inner.clone();
-                let mut partial_outp = vec![F::zero(); outp.len()];
-                inner.call_poly(&mut partial_outp, chunk).unwrap();
-                partial_outp
-            })
+            .fold(
+                || ParallelSumFoldState::new(&self.serial_sum.inner, outp.len()),
+                |mut state, chunk| {
+                    state
+                        .inner
+                        .call_poly(&mut state.partial_output, chunk)
+                        .unwrap();
+                    for (sum_elem, output_elem) in state
+                        .partial_sum
+                        .iter_mut()
+                        .zip(state.partial_output.iter())
```

Comment:
nit: `Iterator::zip` takes `IntoIterator`, which ought to be implemented on `Vec<F>`, so I think you could do:
```suggestion
                        .zip(&state.partial_output)
```

Reply:
I need to retain ownership of `state` for reuse in the next fold operation, so that rules out partial move of its members.